js/modules/loki_file_server_api.js

```diff
@@ -40,6 +43,184 @@ class LokiFileServerAPI {
+                      isRequest
+                        ? textsecure.protobuf.PairingAuthorisationMessage.Type
+                            .REQUEST
+                        : textsecure.protobuf.PairingAuthorisationMessage.Type
+                            .GRANT
+                    );
+                    // log.info('auth is valid for', user.username)
+                    if (iterator(user.username, auth)) {
+                      found = true;
+                    }
+                  } catch (e) {
+                    log.warn(
+                      `Invalid signature on pubkey ${
+                        user.username
+                      } authorization ${
+                        auth.secondaryDevicePubKey
+                      } isRequest ${isRequest}`
+                    );
+                  }
+                }); // end forEach authorisations
+              }
+            }
+          }
+        }); // end forEach annotations
+      }
+      if (notFoundHandler) {
+        if (found) {
+          return;
+        }
+        notFoundHandler(user.username);
+      }
+    }); // end forEach users
+    // log.info('done with users', users.length);
+  }
+
+  // verifies list of pubKeys for any deviceMappings
+  // returns the relevant primary pubKeys
+  async verifyPrimaryPubKeys(pubKeys) {
+    const newSlavePrimaryMap = {}; // new slave to primary map
+    const checkSigs = {}; // cache for authorization
+    const primaryPubKeys = [];
+
+    // go through multiDeviceResults and get primary Pubkey
+    await this.verifyUserObjectDeviceMap(pubKeys, true, (slaveKey, auth) => {
+      // log.info('slave iterator', slaveKey);
+      // if it doesn't throw, that means it's valid
+      // add map to newSlavePrimaryMap
+      if (
+        newSlavePrimaryMap[slaveKey] &&
+        newSlavePrimaryMap[slaveKey] !== auth.primaryDevicePubKey
+      ) {
+        log.warn(
+          `file server user annotation primaryKey mismatch, had ${
+            newSlavePrimaryMap[slaveKey]
+          } now ${auth.primaryDevicePubKey} for ${slaveKey}`
+        );
+        return;
+      }
+      // log.info('valid', slaveKey);
+      if (primaryPubKeys.indexOf(`@${auth.primaryDevicePubKey}`) === -1) {
+        primaryPubKeys.push(`@${auth.primaryDevicePubKey}`);
+      }
+      checkSigs[slaveKey] = auth;
+      newSlavePrimaryMap[slaveKey] = auth.primaryDevicePubKey;
+    }); // end verifyUserObjectDeviceMap
+
+    // log.info('verifyUserObjectDeviceMap', pubKeys, '=>', primaryPubKeys);
+
+    // no valid primary pubkeys to check
+    if (!primaryPubKeys.length) {
+      return [];
+    }
+
+    const verifiedPrimaryPKs = [];
+
+    // get a list of all of primary pubKeys to verify the secondaryDevice assertion
+    await this.verifyUserObjectDeviceMap(
+      primaryPubKeys,
+      false,
+      (primaryKey, auth) => {
+        // log.info('primary iterator', slaveKey);
+        if (verifiedPrimaryPKs.indexOf(`@${primaryKey}`) === -1) {
```

Comment:
Is the `@` really needed here? If so, we can add it only where needed?

Reply:
yea we need to pass @ when it's a username not a userid to the REST API, so instead of looping later and adding it in, I just add it in the beginning.

Reply:
What I meant was that the function here shouldn't need to know about the REST API stuff. There is a bit of duplication while adding an `@` 4 times, 2 times when pushing into the array, then again 2 times when finding a pubkey in the the array.

We should be able to restrict the `@` to one single location in the code, before making the API call:
```js
const verifiedDeviceResults = await this.serverAPI.getUsersAnnotations(
              verifiedPrimaryPKs.map(pk => `@${pk}`)
            );
```

Reply:
a map is still a loop, I'm just preferring less of them by eating some memory higher up.  The function is only there (currently) for the REST API, so I'm not sure the point of making it more general would be.